test/unit/math/test_ad_matvar.hpp

```diff
@@ -0,0 +1,724 @@
+  if (!(sizeof...(T1) == sizeof...(T2))) {
+    FAIL() << "The number of arguments in each tuple must match";
+  }
+
+  expect_near_rel_matvar(message, x, y, tols,
+                         std::make_index_sequence<sizeof...(T1)>());
+}
+///@}
+
+/** @name test_matvar_gradient
+ * Given the inputs and outputs of a function called with Eigen
+ * matrices of vars (matvars) and vars with inner Eigen types (varmats),
+ * check that the values of the outputs are equal and that gradients with
+ * respect to each output are the same
+ *
+ * @tparam ResultMatVar Output type of matvar calculation
+ * @tparam ResultVarMat Output type of varmat calculation
+ * @tparam MatVarArgs Argument types for matvar calculation
+ * @tparam VarMatArgs Argument types for varmat calculation
+ * @param tols Test tolerances
+ * @param A_mv_f Output for matvar calculation
+ * @param A_vm_f Output for varmat calculation
+ * @param A_mv_tuple Input arguments to matvar calculation as tuple
+ * @param A_vm_tuple Input arguments to varmat calculation as tuple
+ */
+///@{
+/**
+ * Overload for scalar outputs
+ */
+template <typename ResultMatVar, typename ResultVarMat, typename... MatVarArgs,
+          typename... VarMatArgs,
+          require_all_var_vt<std::is_arithmetic, ResultMatVar,
+                             ResultVarMat>* = nullptr>
+inline void test_matvar_gradient(const ad_tolerances& tols,
+                                 ResultMatVar& A_mv_f, ResultVarMat& A_vm_f,
+                                 const std::tuple<MatVarArgs...>& A_mv_tuple,
+                                 const std::tuple<VarMatArgs...>& A_vm_tuple) {
+  expect_near_rel("var<Matrix> vs Matrix<var> values", value_of(A_vm_f),
+                  value_of(A_mv_f), tols.gradient_val_);
+  A_vm_f.adj() = 1;
+  A_mv_f.adj() = 1;
+  stan::math::grad();
+  expect_near_rel_matvar("var<Matrix> vs Matrix<var>", A_vm_tuple, A_mv_tuple,
+                         tols);
+  stan::math::set_zero_all_adjoints();
+}
+
+/**
+ * Overload for `std::vector` outputs
+ */
+template <typename ResultMatVar, typename ResultVarMat, typename... MatVarArgs,
+          typename... VarMatArgs,
+          require_std_vector_vt<is_var, ResultMatVar>* = nullptr,
+          require_std_vector_vt<is_var, ResultVarMat>* = nullptr>
+inline void test_matvar_gradient(const ad_tolerances& tols,
+                                 ResultMatVar& A_mv_f, ResultVarMat& A_vm_f,
+                                 const std::tuple<MatVarArgs...>& A_mv_tuple,
+                                 const std::tuple<VarMatArgs...>& A_vm_tuple) {
+  expect_near_rel("var<Matrix> vs Matrix<var> values", value_of(A_vm_f),
+                  value_of(A_mv_f), tols.gradient_val_);
+  for (size_t i = 0; i < A_vm_f.size(); ++i) {
+    A_vm_f[i].adj() = 1;
+    A_mv_f[i].adj() = 1;
+    stan::math::grad();
+    expect_near_rel_matvar("var<Matrix> vs Matrix<var>", A_vm_f[i], A_mv_f[i],
+                           tols);
```

Comment:
Would it be safer to check all of the values here? Just to make sure we are indexing correctly everywhere etc

Reply:
The output values are getting checked out of the loop, and all of them are getting checked I think. Moving the checks in the loop just assumes the inputs and outputs are the same lengths (which they probably are but not necessary)

This line checks the input values and adjoints.